### ISSUE TITLE
Support negative indices in string slices

### ISSUE THREAD

**Issue by Crestwave** [oilshell/oil#772]
I'm not sure if this is an intentional incompatibility since there's a specific warning about it (that I stole for the issue title). After all, the conflict between `: -1` and `:-1` is a bit weird, but `nosj` uses it. and the workaround I came up with is a bit wordy (`"${foo:$((${#foo}?${#foo}-1:0))}"`)

```bash
$ foo=bar
$ echo "${foo: -1}"
r
```

```bash
osh$ foo=bar
osh$ echo "${foo: -1}"
  echo "${foo: -1}"
        ^~
[ interactive ]:2: warning: The start index of a string slice can't be negative: -1
```

**Comment by andychu** [comment]
Yes I ran into this in mal Lisp too.  Looks like we should fix it.

I think the hangup was that negative indices and UTF-8 are not trivial, but I think we have a routine for it now

**Comment by andychu** [comment]
@Crestwave OK I implemented this and took out the patch for mal Lisp as well.

I should submit that patch upstream -- it's pretty minimal and improves the code IMO (quoting for associative array keys, don't confuse strings and arrays, and extract variable for a regex which simplifies it)

Implemented using UTF-8 PreviousChar routine by @asokoloski   :)
